odoo/addons/base/models/qweb.py

```diff
@@ -397,397 +246,340 @@ def get_template(self, template, options):
-                if hasattr(self, interpret_handler):
-                    _logger.warning(
-                        "Directive '%s' must be AST-compiled. Dynamic interpreter %s will ignored",
-                        mname, interpret_handler
-                    )
+        return []
+
+    def _compile_options(self, el, varname, options, indent):
+        """
+        compile t-options and add to the dict the t-options-xxx values
+        """
+        t_options = el.attrib.pop('t-options', None)
+        directive = ''
+        if t_options:
+            directive = f't-options="{t_options}"'
+
+        code = []
+        dict_arg = []
+        for key in list(el.attrib):
+            if key.startswith('t-options-'):
+                value = el.attrib.pop(key)
+                directive += f' {key}={repr(value)}'
+                option_name = key[10:]
+                dict_arg.append(f'{repr(option_name)}:{self._compile_expr(value)}')
+
+        if t_options and dict_arg:
+            code.append(self._indent(f"{varname} = {{**{self._compile_expr(t_options)}, {', '.join(dict_arg)}}}", indent))
+        elif dict_arg:
+            code.append(self._indent(f"{varname} = {{{', '.join(dict_arg)}}}", indent))
+        elif t_options:
+            code.append(self._indent(f"{varname} = {self._compile_expr(t_options)}", indent))
+
+        return code
+
+    def _compile_format(self, expr):
+        """ Parses the provided format string and compiles it to a single
+        expression python, uses string with format method.
+        Use format is faster to concat string and values.
+        """
+        text = ''
+        values = []
+        base_idx = 0
+        for m in _FORMAT_REGEX.finditer(expr):
+            literal = expr[base_idx:m.start()]
+            if literal:
+                text += literal.replace('{', '{{').replace("}", "}}")
+            text += '{}'
+            values.append(f'self._compile_to_str({self._compile_expr(m.group(1) or m.group(2))})')
+            base_idx = m.end()
+        # string past last regex match
+        literal = expr[base_idx:]
+        if literal:
+            text += literal.replace('{', '{{').replace("}", "}}")
+
+        code = repr(text)
+        if values:
+            code += f'.format({", ".join(values)})'
+        return code
+
+    def __compile_expr_tokens(self, tokens, allowed_keys, raise_on_missing=False):
+        """Transform the list of token coming into a python instruction in
+        textual form by adding the namepaces for the dynamic values.
+
+        Example: `5 + a + b.c` to be `5 + values.get('a') + values['b'].c`
+        Unknown values are considered to be None, but using `values['b']`
+        gives a clear error message in cases where there is an attribute for
+        example (have a `KeyError: 'b'`, instead of `AttributeError: 'NoneType'
+        object has no attribute 'c'`).
+        """
+        # found and extract allowed values
```

Comment:
This doesn't really tell the reader why what's happening, which is what I was looking for e.g. 

> Finds and extracts the current "scope"'s "allowed values": values which should not be accessed through the environment's namespace:
>
> * the local variables of a lambda should be accessed directly e.g. `lambda a: a + b` should be compiled to `lambda a: a + values['b']`, since `a` is local to the lambda it has to be accessed directly but `b` needs to be accessed through the rendering environment
> * similarly for a comprehensions `[a + b for a in c]` should be compiled to `[a + values.get('b') for a in values.get('c')]`
>
> to avoid the risk of confusion between nested lambdas / comprehensions, this is currently performed independently at each level of brackets nesting (hence the function being recursive).